abi-versions/vNEXT/README.md

```diff
@@ -0,0 +1,1471 @@
+
+#### `proxy_dispatch_http_call`
+
+* params:
+  - `i32 (const char*) upstream_name_data`
+  - `i32 (size_t) upstream_name_size`
+  - `i32 (iovec_array_t[]) headers_map_data`
+  - `i32 (size_t) headers_map_size`
+  - `i32 (iovec_array_t) body_data`
+  - `i32 (size_t) body_size`
+  - `i32 (iovec_array_t[]) trailers_map_data`
+  - `i32 (size_t) trailers_map_size`
+  - `i32 (uint32_t) timeout_milliseconds`
+  - `i32 (uint32_t*) return_callout_id`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Dispatches a HTTP call to upstream (`upstream_name_data`, `upstream_name_size`).
+
+Once the response is returned to the host, `proxy_on_http_call_response` will be
+called with an unique call identifier (`return_callout_id`).
+
+Returns `Ok` on success.
+
+Returns `NotFound` for unknown upstream.
+
+Returns `NotAllowed` when requests to given upstream are not allowed.
+
+Returns `InvalidMemoryAccess` when `upstream_name_data`, `upstream_name_size`,
+`headers_map_data`, `headers_map_size`, `body_data`, `body_size`,
+`trailers_map_data`, `trailers_map_size` and/or `return_callout_id` point to
+invalid memory address.
+
+### Callbacks optionally exposed by the Wasm module
+
+#### `proxy_on_http_call_response`
+
+* params:
+  - `i32 (uint32_t) callout_id`
+  - `i32 (size_t) num_headers`
+  - `i32 (size_t) body_size`
+  - `i32 (size_t) num_trailers`
+* returns:
+  - none
+
+Called when the complete response to the HTTP call (`callout_id`) is received.
+
+Headers (`num_headers` entries) and trailers (`num_trailers` entries) can be
+retrieved using `proxy_get_map_values`.
+
+Response body (of size `body_size`) can be retrieved using `proxy_get_buffer`.
+
+
+## gRPC callouts
+
+### Functions exposed by the host environment
+
+#### `proxy_dispatch_grpc_call`
+
+* params:
+  - `i32 (const char*) upstream_name_data`
+  - `i32 (size_t) upstream_name_size`
+  - `i32 (const char*) service_name_data`
+  - `i32 (size_t) service_name_size`
+  - `i32 (const char*) service_method_data`
+  - `i32 (size_t) service_method_size`
+  - `i32 (iovec_array_t[]) initial_metadata_map_data`
+  - `i32 (size_t) initial_metadata_map_size`
+  - `i32 (iovec_array_t) grpc_message_data`
+  - `i32 (size_t) grpc_message_size`
+  - `i32 (uint32_t) timeout_milliseconds`
+  - `i32 (uint32_t*) return_callout_id`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Dispatch a call to a gRPC service (`service_name_data`, `service_name_size`)
+using upstream (`upstream_name_data`, `upstream_name_size`).
+
+Once the response is returned to the host, `proxy_on_grpc_call_response` and/or
+`proxy_on_grpc_call_close` will be called with an unique call identifier
+(`return_callout_id`). The unique call identifier can also be used to cancel
+outstanding requests using `proxy_cancel_grpc_call` or close outstanding request
+using or `proxy_close_grpc_call`.
+
+Returns `Ok` on success.
+
+Returns `NotFound` for unknown service (`service_name_data`,
+`service_name_size`).
```

Comment:
Doesn't that error come back from the upstream gRPC server? In that case would the error really be there or in the response? OTOH, I imagine that an invalid upstream name would immediately return an error.